src/test/java/com/microsoft/azure/relay/SendReceiveTest.java

```diff
@@ -304,38 +296,50 @@ private static void httpRequestSender(String method, String msgExpected, String
 		out.flush();
 		out.close();
 
-		String inputLine;
-		StringBuilder builder = new StringBuilder();
-		BufferedReader inStream = new BufferedReader(new InputStreamReader(conn.getInputStream()));
-		
+		String received = readStringFromStream(conn.getInputStream());
 		assertEquals("Http connection sender did not receive the expected response code.", STATUS_CODE, conn.getResponseCode());
 		assertEquals("Http connection sender did not receive the expected response description.", STATUS_DESCRIPTION, conn.getResponseMessage());
-		
-		while ((inputLine = inStream.readLine()) != null) {
-			builder.append(inputLine);
-		}
-		inStream.close();
 		assertEquals("Http connection sender did not receive the expected response message size.", 
 				msgExpected.length(), 
-				builder.toString().length());
-		assertEquals("Http connection sender did not receive the expected response message.", msgExpected, builder.toString());
+				received.length());
+		assertEquals("Http connection sender did not receive the expected response message.", msgExpected, received);
 	}
 	
-	private static void httpRequestHandlerMultipleSend(RelayedHttpListenerContext context, String firstMsg, String secondMsg, boolean hasPause) {
+	private static void sendResponseMessages(RelayedHttpListenerContext context, String[] messages, boolean hasPause) {
 		RelayedHttpListenerResponse response = context.getResponse();
 		response.setStatusCode(STATUS_CODE);
 		response.setStatusDescription(STATUS_DESCRIPTION);
 		
+		CompletableFuture<?>[] cfs = new CompletableFuture<?>[messages.length];
 		try {
-			response.getOutputStream().write(firstMsg.getBytes());
-			if (hasPause) {
-				Thread.sleep(2500);
+			for (int i = 0; i < messages.length; i++) {
+				cfs[i] = response.getOutputStream().writeAsync(messages[i].getBytes(), 0, messages[i].length());
+				
+				// Pause for testing the flush timeout
+				if (hasPause && i < messages.length - 1) {
+					Thread.sleep(ResponseStream.WRITE_BUFFER_FLUSH_TIMEOUT_MILLIS);
+				}
 			}
-			response.getOutputStream().write(secondMsg.getBytes());
-		} catch (IOException | InterruptedException e) {
+			CompletableFuture.allOf(cfs).join();
+		} catch (Exception e) {
 			fail(e.getMessage());
 		} finally {
 		    context.getResponse().close();
 		}
 	}
+	
+	private static String readStringFromStream(InputStream inputStream) throws IOException {
+		if (inputStream == null) {
+			return emptyStr;
+		}
+		
+		StringBuilder builder = new StringBuilder();
+		String inputLine;
+		try (BufferedReader inStream = new BufferedReader(new InputStreamReader(inputStream))){
+			while ((inputLine = inStream.readLine()) != null) {
```

Comment:
>readLine [](start = 32, length = 8)

I feel like I've asked this before. What happens with multi-line strings? Does '\n' get removed from the inputStream in this process? What about '\r\n' in the original data? I hope we have some tests which are using boring-old bytes to make sure the SDK isn't messing with the user data in any way at all.

Reply:
I will check on that in a separate PR

---
In reply to: [293099146](https://github.com/Azure/azure-relay-java/pull/46#discussion_r293099146) [](ancestors = 293099146)